shapeshift/web
# feat: wire-up Ledger hdwallet

## Description

Working in this base functionality PR:

- [x] Receive
- [x] Sends
- [x] Txs
- [x] Swaps - with caveats, only same-chain for now until we cache things properly
- [x] Re-pair device / race `'TransportRaceCondition: An action was already pending on the Ledger device. Please deny or reconnect.'` condition
- [x] Reconnect on app close disconnect i.e `DisconnectedDeviceDuringOperation: Failed to execute 'transferOut' on 'USBDevice': The device was disconnected.`

To be implemented in follow-ups:

- Ledger support for Cosmos and THORChain (RUNE)
- BCH receive address format - fixed in https://github.com/shapeshift/hdwallet/pull/640
- Accounts caching and/or UX improvements WRT the need to open xyz app first
- Verification of addresses at key times so users can ensure they're dealing with the right Ledger/account
- Programatically open the relevant app before `call()` to specific apps?
- Cross-chain swaps - can't be implemented until either/all 3 above are implemented
## Pull Request Type


## Issue (if applicable)

N/A

## Risk



Large regression risk in `useReceiveAddress` / `getReceiveAddress` being too strongly memoized i.e stale receive address on wallet change/"update" for all wallets. Think things like next receive address being broken for UTXOs, receive addresses when switching wallets or MM account being the one from the previously connected wallet/account, etc.

## Testing



- Ledger pairing works
- Ledger chains connection works (currently requires to open the matching app for each chain before clicking "Connect" or doing any operation i.e Tx signing or receive address for a given chain
- Sends work for all chains (assumes the relevant app is currently open before starting the send flow)
  - Make sure to test with Legacy/SegWit/SegWit Native for LTC and BTC
- Accounts are loaded for all chains and script types i.e all EVM chains, DOGE, BCH, BTC (all 3 script types) and LTC (all 3 script types)
- Ensure the accounts are matching the ones from native - import the same seed as native into your Ledger for testing

### Engineering



- All Ledger capability is safely flagged off

### Operations



- ☝🏽 

## Screenshots (if applicable)

<img width="477" alt="image" src="https://github.com/shapeshift/web/assets/17035424/60b65565-0c00-4a31-b6b6-f7196549a221">
<img width="476" alt="image" src="https://github.com/shapeshift/web/assets/17035424/d8c09a1b-64cd-4ed2-ba56-406e0baef184">
<img width="1020" alt="image" src="https://github.com/shapeshift/web/assets/17035424/5c172dc0-1346-4412-8519-f6b22eb63eb8">

https://github.com/shapeshift/web/assets/17035424/f70c59d0-6ed1-450d-9913-cd31a56e9060

<img width="1418" alt="image" src="https://github.com/shapeshift/web/assets/17035424/43aa9027-1d18-4fcc-910b-efe150112e8d">
<img width="1413" alt="image" src="https://github.com/shapeshift/web/assets/17035424/f79f0817-69ea-4954-adc2-8d11633bf6e7">
<img width="489" alt="image" src="https://github.com/shapeshift/web/assets/17035424/320bb080-9e1d-48aa-93f6-39f674f17d80">
<img width="1411" alt="image" src="https://github.com/shapeshift/web/assets/17035424/93f3aee6-8e3a-4e72-8289-9d37863050be">
<img width="476" alt="image" src="https://github.com/shapeshift/web/assets/17035424/5242362e-c83f-4368-bce1-eff0db353c05">
<img width="1409" alt="image" src="https://github.com/shapeshift/web/assets/17035424/f4eb6a6d-762e-4db5-9806-f111f4ab308b">
<img width="496" alt="image" src="https://github.com/shapeshift/web/assets/17035424/7b1acb21-e053-4355-81b5-e5410f47bc43">
<img width="955" alt="image" src="https://github.com/shapeshift/web/assets/17035424/8e2d5ebe-00a7-4802-8f66-598c3053837f">


Note the non-standard receive address for BCH here - to be tackled in a follow-up. It does work (tested with self sends as can be seen above), although does shows up with a warning WRT unverfied inputs on the Ledger currently, which may be related to the address format.

<img width="503" alt="image" src="https://github.com/shapeshift/web/assets/17035424/7dffc686-9a05-4048-8afc-c13d016a2c99">
<img width="1412" alt="image" src="https://github.com/shapeshift/web/assets/17035424/5676782d-7a31-40d6-9010-49e09a524467">
